### PR TITLE
NEW :  Uniformize member stats in boxes and index

### PR DESCRIPTION
**Pull request by daraelmin**
One request per graph instead of four or five

Befor PR

image

After PR

image

### CODE REVIEW

File: htdocs/adherents/class/adherentstats.class.php [hunk 1]
```diff
@@ -181,4 +182,56 @@ public function getAllByYear()
...
+	 * 	@return		array                   Array with
+	 */
+	public function countMembersByTypeAndStatus()
+	{
+		global $user;
+
+		$now = dol_now();
+
+		$sql = "SELECT COALESCE(d.fk_adherent_type, 'total') as fk_adherent_type, t.libelle as label";
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_DRAFT." THEN 'members_draft' ELSE NULL END) as members_draft";
```
Comment (eldy) [inline htdocs/adherents/class/adherentstats.class.php:199]: Do not use CASE WHEN into sql, use instead  $db->ifsql method that will generate the sql using the if with the syntax according to database manager.

File: htdocs/adherents/class/adherentstats.class.php [hunk 2]
```diff
@@ -181,4 +182,56 @@ public function getAllByYear()
...
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_VALIDATED."  AND (d.datefin IS NULL AND t.subscription = '1') THEN 'members_pending' ELSE NULL END) as members_pending";
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_VALIDATED."  AND (d.datefin >= '".$this->db->idate($now)."' OR t.subscription = 0) THEN 'members_uptodate' ELSE NULL END) as members_uptodate";
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_VALIDATED."  AND (d.datefin < '".$this->db->idate($now)."' AND t.subscription = 1) THEN 'members_expired' ELSE NULL END) as members_expired";
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_EXCLUDED." THEN 'members_excluded' ELSE NULL END) as members_excluded";
+		$sql .= ", COUNT(CASE WHEN d.statut = ".Adherent::STATUS_RESILIATED."  THEN 'members_resiliated' ELSE NULL END) as members_resiliated";
+		$sql .= " FROM ".MAIN_DB_PREFIX."adherent as d , ".MAIN_DB_PREFIX."adherent_type as t";
```
Comment (eldy) [inline htdocs/adherents/class/adherentstats.class.php:205]: If a type has no member at all, it seems this request will return nothing when need need a count per type, so if there is no member at all for this type, we need a result into the sql request. May be we need to convert the join into left join ?

File: htdocs/adherents/class/adherentstats.class.php [hunk 3]
```diff
@@ -181,4 +182,56 @@ public function getAllByYear()
...
+		$sql .= " WHERE t.rowid = d.fk_adherent_type";
+		$sql .= " AND d.entity IN (" . getEntity('adherent') . ")";
+		$sql .= " AND t.entity IN (".getEntity('member_type').")";
+		$sql .= " AND t.statut = 1";
+		$sql .= " GROUP BY d.fk_adherent_type";
+		$sql .= " WITH ROLLUP;";
```
Comment (eldy) [inline htdocs/adherents/class/adherentstats.class.php:211]: This instruction must be avoided :  Not managed the same way by different databases.
Also the use of ROLLUP break the purity of data. Intermediary aggregation can easily be done using PHP instead keeping the array of data no corrupted by artificially inserted data and keep the array result pure.

File: htdocs/adherents/class/adherentstats.class.php (hunk 1 above)
Comment (daraelmin) [reply htdocs/adherents/class/adherentstats.class.php]: Done

File: htdocs/adherents/class/adherentstats.class.php (hunk 2 above)
Comment (daraelmin) [reply htdocs/adherents/class/adherentstats.class.php]: Done

File: htdocs/adherents/class/adherentstats.class.php (hunk 3 above)
Comment (daraelmin) [reply htdocs/adherents/class/adherentstats.class.php]: Done

**Comment by daraelmin** [comment]
Fixed as suggested
Add a new column "Total by status" in box
Change label to match the graph and the left menu

**Comment by daraelmin** [comment]
@eldy hope it's better now

**Comment by daraelmin** [comment]
With all last modification (Add total, limit in time, and add period in title)

image

**Pull request closed without merging**